pkg/addon/source.go

```diff
@@ -392,23 +150,63 @@ func NewAsyncReader(baseURL, bucket, subPath, token string, rdType ReaderType) (
 			bucketEndPoint = fmt.Sprintf(bucketTmpl, ossURL.Scheme, bucket, ossURL.Host)
 		}
 		return &ossReader{
-			baseReader:     bReader,
 			bucketEndPoint: bucketEndPoint,
 			path:           subPath,
 			client:         resty.New(),
 		}, nil
 	}
-	return nil, errors.New("addon registry invalid")
+	return nil, fmt.Errorf("invalid addon registry type '%s'", rdType)
+}
+
+// BuildReader will build a AsyncReader from registry, AsyncReader are needed to read addon files
+func (r Registry) BuildReader() (AsyncReader, error) {
+	if r.OSS != nil {
+		o := r.OSS
+		return NewAsyncReader(o.Endpoint, o.Bucket, o.Path, "", ossType)
+	} else if r.Git != nil {
+		g := r.Git
+		return NewAsyncReader(g.URL, "", g.Path, g.Token, gitType)
+	}
+	return nil, errors.New("registry don't have enough info to build a reader")
+
+}
+
+// GetUIData get UIData of an addon
+func (r Registry) GetUIData(meta *SourceMeta, opt ListOptions) (*UIData, error) {
```

Comment:
It seems better if these function owned by the AsyncReader, you shouldn't create a reader every time you revoke this function.

And the same below.

Reply:
AsyncReader are limited to read files from OSS/GitHub. This interface is not about addon business